src/js_native_api_types.h

```diff
@@ -22,6 +22,33 @@ typedef uint16_t char16_t;
 // JSVM API types are all opaque pointers for ABI stability
 // typedef undefined structs instead of void* for compile time type safety
 typedef struct napi_env__* napi_env;
+
+// We need to mark APIs which can be called during garbage collection (GC),
+// meaning that they do not affect the state of the JS engine, and can
+// therefore be called synchronously from a finalizer that itself runs
+// synchronously during GC. Such APIs can receive either a `napi_env` or a
+// `node_api_nogc_env` as their first parameter, because we should be able to
+// also call them during normal, non-garbage-collecting operations, whereas
+// APIs that affect the state of the JS engine can only receive a `napi_env` as
+// their first parameter, because we must not call them during GC. In lieu of
+// inheritance, we use the properties of the const qualifier to accomplish
+// this, because both a const and a non-const value can be passed to an API
+// expecting a const value, but only a non-const value can be passed to an API
+// expecting a non-const value.
+//
+// In conjunction with appropriate CFLAGS to warn us if we're passing a const
+// (nogc) environment into an API that expects a non-const environment, and the
+// definition of nogc finalizer function pointer types below, which receive a
+// nogc environment as their first parameter, and can thus only call nogc APIs
+// (unless the user explicitly casts the environment), we achieve the ability
+// to ensure at compile time that we do not call APIs that affect the state of
+// the JS engine from a synchronous (nogc) finalizer.
+#if defined(NAPI_EXPERIMENTAL) && defined(NODE_API_EXPERIMENTAL_NOGC_ENV)
```

Comment:
Can we avoid adding the experimental feature flags like `NODE_API_EXPERIMENTAL_NOGC_ENV`?
It would be better just to use experimental version.
It is virtually impossible to write good unit tests that exercise all possible combinations of experimental feature flags.
I believe the experimental features are unnecessary increase the complexity of code, tests, docs, etc.
Currently adding new or experimental version already covers adding new APIs and changing API behavior.

Reply:
Let's review our decision to use the secondary experimental flag at our next Node-API meeting.

Reply:
Actually, I just came across a good reason to always have a second flag within `NAPI_EXPERIMENTAL`:
I turned on `NAPI_EXPERIMENTAL` in order to update node-addon-api to use `node_api_post_finalizer`, otherwise finalizers will crash when running against v21.1.0 or later. So, basically, inside napi-inl.h,
```cpp
#ifdef NAPI_EXPERIMENTAL
napi_status status = node_api_post_finalizer(env, data, hint, finalizer);
...
#else
finalizer(env, data, hint);
#endif
```
This works great ... _**if**_ you build against 21.1.0 or later. However, if you build against v18.x, it will fail, because `node_api_post_finalizer` was not available in v18.x, not even with `NAPI_EXPERIMENTAL` defined.

So, basically, the meaning of `NAPI_EXPERIMENTAL` depends on the Node.js version. For example, in v18.x, `NAPI_EXPERIMENTAL` means that `node_api_symbol_for` is available, because back then it was not yet marked as stable. In v21.1.0, `node_api_symbol_for` is stable, and `NAPI_EXPERIMENTAL` means something completely different, namely that `node_api_post_finalizer` is available.

Thus, `NAPI_EXPERIMENTAL` doesn't tell us what additional things are available, because that list depends on the Node.js version. So, we should _**always**_ have a second flag that can only be used in conjunction with `NAPI_EXPERIMENTAL`, that is experimental-feature-specific. That way, if we build with `NAPI_EXPERIMENTAL` but against an old Node.js version, we will know the feature is unavailable because the second flag will not be defined.

I guess we should probably make a PR that adds:
```diff
 #ifdef NAPI_EXPERIMENTAL

+#define NODE_API_HAVE_POST_FINALIZER
 NAPI_EXTERN napi_status NAPI_CDECL
 node_api_post_finalizer(napi_env env,
                         napi_finalize finalize_cb,
                         void* finalize_data,
 
```

so that we can then do
```cpp
#if defined(NAPI_EXPERIMENTAL) && defined(NODE_API_HAVE_POST_FINALIZER)
napi_status status = node_api_post_finalizer(env, data, hint, finalizer);
...
#else
finalizer(env, data, hint);
#endif
```

We may want to update our new Node-API policy to include the requirement that a corresponding `NODE_API_HAVE_<FEATURE_NAME>` always be defined along with the new feature.